Fix retain cycles preventing objects from being deallocated after logging out

staskus opened the pull request:
Fixes #21010

Memory is not deallocated after logging out.

It is caused by 2 main issues:
- `WPTabBarController` held strongly by singleton `RootViewCoordinator` after logging out ([see more details](https://github.com/wordpress-mobile/WordPress-iOS/issues/21010#issuecomment-1623846522))
- Dashboard, Reader, Notification, and other initial view controllers and dependencies that have retain cycles preventing them from releasing ([see more details](https://github.com/wordpress-mobile/WordPress-iOS/issues/21010#issuecomment-1628953900))

## Solution

To address issues in `RootViewCoordinator` I decided to limit the changes as much as possible to avoid cascading code changes throughout the codebase. For now, I decided to manually release a reference to `RootViewPresenter` (`WPTabBarController`) once logging out. See more info in commit message.

## ⚠️ Other issues not addressed by this PR

- I noticed that the memory footprint remains high after logout, it may be more related to stores or services, I need to investigate more
- There's a significant number of memory leaks in the app. This login -> logout case helps to quickly reproduce and confirm whether the issues with initial VCs are fixed but I think we'll need to create more tasks to go through all the main flows
- There looks to be a memory leak with Login VCs as well, they are allocated and not deallocated when logging in and logging out as well https://github.com/wordpress-mobile/WordPress-iOS/issues/21075

## To test:

### Login & Logout (On both Jetpack and WordPress apps)
1. Login
2. Logout
3 Check Memory Graph
4. WPTabBarController and other non-login related view controllers should be deallocated

### Regression: WordPress without static posters
1. Open WordPress
2. Enabled all the phases
3. Relaunch application
4. Confirm that no tabs are displayed and navigation in menu works

### Regression: WordPress with static posters
1. Open WordPress
2. Enable static posters phase
3. Relaunch application
4. Confirm that tabs are displayed but Reader and Notifications tab is showing static poster

### Regression: Dashboard Blaze card
1. Open Jetpack
2. Log into the .com site with a post
3. Select "Promote your content with Blaze" card
6. Confirm that navigation to Blaze controller is working

### Regression: Dashboard Quick Actions
1. Open Jetpack
2. Log into the account
3. Select "Stats, "Posts", "Pages", "Media" buttons
4. Confirm that navigation works

### Regression: Dashboard navigation
1. Open Jetpack
2. Confirm site picker, home, and menu navigation works

### Regression: Reader navigation
1. Open Jetpack
2. Select Reader tab
3. Confirm opening Post works
4. Confirm opening Search works

## Regression Notes
1. Potential unintended areas of impact

Breaking the currently existing functionality, especially root navigation

7. What I did to test those areas of impact (or what existing automated tests I relied on)

- Unit test suite
- UI test suite
- Manual testing

8. What automated tests I added (or what prevented me from doing so)

None for now but It may be worth considering about performance tests or tests confirming the deallocation of main VCs

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Set weak reference for BlogDashboardViewController in DashboardBlazeCardCell

staskus committed:
Weakly capture action closure in DashbordPagesListCardCell

staskus committed:
Pass weak closure instead of strongly captured method in DashboardPromptCardCell

staskus committed:
Break circular dependency cycle for BlogDashboardViewController in quick action cell

staskus committed:
Break circular dependency cycle for BlogDashboardViewController in DashboardStatsCardCell

staskus committed:
Capture BlogDashboardViewController weekly in BlogDashboardViewModel

staskus committed:
Break retain cycle by capturing presented view controllers weakly in MySiteViewController

staskus committed:
Weak self in NotificationsViewController

staskus committed:
Capture ReaderTabViewModel weakly in a closure

staskus committed:
Use weak closures instead of passing methods to a closure to avoid retain cycle

staskus commented:
@guarani Just informing you that it's still in progress, although most of the classes are finally released after the logout:

- The work still requires thorough testing since it involves architectural changes at `RootViewCoordinator` 
- There's a significant number of memory leaks in the app. This login -> logout case helps to quickly reproduce and confirm whether the issues are fixed but I think we'll need to create more tasks to go through all the main flows
- Even when dozens of VCs, Views, and related classes are deallocated after logout, the memory footprint in the app still didn't decrease. I haven't dug deep but it feels like some blog or site-related information is cached and kept in memory after logout. Our Stores and some services are either singletons or held by singletons which doesn't allow them to be released automatically after logout

staskus committed:
Merge branch 'trunk' into fix/21010-memory-not-deallocated-after-logging-out

staskus committed:
Resolve memory leaks in RootViewCoordinator

Problem:
- RootViewCoordinator is a singleton
- RootViewCoordinator holds a strong reference to RootViewPresenter
- WPTabBarController implements RootViewPresenter and is presenter as rootViewController
- When WPTabBarController is dismissed, RootViewCoordinator continues to hold a strong reference to a WPTabBarController creating a memory leak

Solution:
- Move presentation logic from WindowsManager to RootViewCoordinator
- Only initialize RootViewPresenter when needed
- Release RootViewPresenter after logging out. Not making RootViewPresenter weak or unowned to avoid further cascading changes through the codebase to handle RootViewPresenter being optional
- RootViewPresenter should not be accessed before the root view is presented. If it happens - print a warning. All the issues should be addressed with further improvements.

staskus committed:
Update RELEASE-NOTES.txt

guarani commented on RELEASE-NOTES.txt:
```diff
@@ -1,3 +1,7 @@
+22.9
+-----
+* [**] [internal] Fix multiple memory leaks after logging in and logging out. [#21047]
```
I think users would be happy to hear about these improvements. Wdyt about a public release note for this? It could be called performance improvements

staskus replied:
Could we do both internal and external? I want it to be visible that I was making changes in the logging-in and logging-out flows but at the same time yes, we could also inform users about us making performance or stability improvements.

guarani replied:
Yeah, I think that could work 👍

guarani commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/Prompts/DashboardPromptsCardCell.swift:
```diff
@@ -289,6 +289,26 @@ class DashboardPromptsCardCell: UICollectionViewCell, Reusable {
 
     // Defines the structure of the contextual menu items.
     private var contextMenuItems: [[MenuItem]] {
+        let viewMoreMenuTapped = { [weak self] in
+            guard let self else { return }
+            self.viewMoreMenuTapped()
+        }
+
+        let skipMenuTapped = { [weak self] in
+            guard let self else { return }
+            self.skipMenuTapped()
+        }
+
+        let learnMoreTapped = { [weak self] in
+            guard let self else { return }
+            self.learnMoreTapped()
+        }
+
+        let removeMenuTapped = { [weak self] in
+            guard let self else { return }
+            self.removeMenuTapped()
+        }
+
```
I guess there's a retain cycle here but it's hard to see. 
1. The user logs out
2. The window removes its strong reference to the dashboard and which strongly held the table view cell
3. The cell holds a strong reference to the `UIButton` and the button holds a strong reference to the cell via the button menu's action items 

So the issue is that the cell and the button have a retain cycle, right?

staskus replied:
> So the issue is that the cell and the button have a retain cycle, right?

Yes. `DashboardPromptCardCell` holds `BlogDashboardCardFrameView` which holds menu actions, that capture `DashboardPromptCardCell` strongly in a closure, creating a reference cycle.

Here's the memory graph after logging out if this change hadn't been made. You can see that `DashboardPromptCardCell` holds the reference to `BlogDashboardCardFrameView` and when we go through all the reference graph we can again reach `DashboardPromptCardCell` which is held strongly by a Swift closure.

https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/ac44cf1c-915e-40de-a6ce-335e5f47b4dc

staskus replied:
The lesson here is to avoid assigning functions to closure as a parameter. It is easy to mismanage memory this way https://www.marisibrothers.com/2017/04/memory-leak-in-swift-assigning-function.html

guarani replied:
Until now I've thought that assigning a function to a callback closure was neat, but TIL it creates a retain cycle 😬 Thanks for catching this.

kean replied:
It is neat 😁
It's usually OK if you are assigning a ViewModel's method:

```swift
self.button.action = self.viewModel.buttonTapped
```

The views are expected to retain their ViewModels.

staskus replied:
Yes, that's correct. However, when we don't have a clear architecture in most of the places in the code, we need to be careful with this.

guarani commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/ViewModel/BlogDashboardViewModel.swift:
```diff
@@ -49,7 +49,7 @@ class BlogDashboardViewModel {
         }
 
-        return DashboardDataSource(collectionView: viewController.collectionView) { [unowned self] collectionView, indexPath, item in
+        return DashboardDataSource(collectionView: viewController.collectionView) { [unowned self, unowned viewController] collectionView, indexPath, item in
```
Even though `viewController` is a weak property, is it a strong reference here due to the local variable? I need to brush up on this!

staskus replied:
I admit, these ones were the hardest for me to understand too. 

This is the cycle we get without this fix:
<img width="1397" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/48cc13d3-d3c8-4944-a843-98bc2a5d6538">

Yes, `BlogDashboardViewModel` is holding a weak reference to `BlogDashboardViewController`:

```
final class BlogDashboardViewModel {
    private weak var viewController: BlogDashboardViewController?
```

However, when we capture  `BlogDashboardViewController` in a  `DashboardDataSource` closure, we capture it strongly. It creates a retain cycle of:
`BlogDashboardViewModel` -> `DashboardDataSource` -> `closure` -> `BlogDashboardViewController` -> `BlogDashboardViewModel`

Here's an isolated example you could run on playground:

```
class DataSource {
    let closure: () -> ()
    
    init(closure: @escaping () -> ()) {
        self.closure = closure
    }
}

class ViewModel {
    weak var viewController: ViewController?
    
    lazy var dataSource: DataSource? = {
        guard let viewController = viewController else {
            return nil
        }
        
        return DataSource {
            print(viewController)
        }
    }()
    
    deinit {
        print("ViewModel deinit")
    }
}

class ViewController {
    let viewModel: ViewModel
    
    init(viewModel: ViewModel) {
        self.viewModel = viewModel
        self.viewModel.viewController = self
        self.viewModel.dataSource?.closure()
    }
    
    deinit {
        print("ViewController deinit")
    }
}

ViewController(viewModel: ViewModel())
```

`ViewController` and `ViewModel` do not print deinit message until we do:

```
        return DataSource { [unowned viewController] in
            print(viewController)
        }
```

guarani replied:
Makes sense now. In the sample code above the retain cycle is due to the strong references:
`ViewController` > `ViewModel` > `DataSource` > `closure` > `ViewController`
So it hasn't got anything to do with the local variable.

guarani reviewed (commented):
I only did a partial review since you mention this isn't ready yet. I just went through the code changes but didn't test it. Once it's ready I can dive deeper and actually test the changes, let me know if this works for you.

I left some comments with first impressions, but they're things I might be able to answer myself after spending more time review and testing.

staskus committed:
Merge branch 'trunk' into fix/21010-memory-not-deallocated-after-logging-out

staskus committed:
Update RELEASE-NOTES.txt

staskus committed:
Added unit tests confirming deallocation of rootViewController after logout

- Made WordPressAuthenticator injectable
- Checking if rootViewController is deinitialized after presenting signInUI

staskus committed:
Merge branch 'trunk' into fix/21010-memory-not-deallocated-after-logging-out

guarani commented on RELEASE-NOTES.txt:
```diff
@@ -1,7 +1,9 @@
+* [*] Performance improvements in login and logouts flows. [#21047]
 * [**] Block editor: Move undo/redo buttons to the navigation bar. [#20930]
-* [*] Fixed an issue that caused the UI to be briefly unresponsive in certian case when opening the app. [#21065]
+* [*] Fixed an issue that caused the UI to be briefly unresponsive in certain case when opening the app. [#21065]
```
This might be here by accident since https://github.com/wordpress-mobile/WordPress-iOS/pull/21065 is already merged to `trunk`.

staskus replied:
Yes, it's already merged. I just fixed a typo `certian` -> `certain`

guarani commented on WordPress/Classes/System/RootViewCoordinator.swift:
```diff
@@ -34,33 +45,58 @@ class RootViewCoordinator {
+
+        windowManager?.show(loginViewController, completion: completion)
+        wordPressAuthenticator.track(.openedLogin)
+        self.rootViewPresenter = nil
```
This looks to be the core of the fix here. I think it will deallocate anything `WPTabBarController` and beyond once we fix the other retain cycles we've talked about 👍

staskus replied:
Yes, it's a core fix. However, if there're retain cycles in other view controllers (and there were plenty), `WPTabBarController` gets deallocated but other ViewControllers and the resources they hold - don't.

guarani commented on WordPress/WordPressTest/RootViewCoordinatorTests.swift:
```diff
@@ -0,0 +1,54 @@
+        sut = nil
+    }
+
+    func testAppUIDeallocatedAfterLogout() {
```
Nice test! It's good to be able to test that things are deallocated after logout.

guarani reviewed (commented):
> ### Login & Logout (On both Jetpack and WordPress apps)
> 1. Login
> 2. Logout
>    3 Check Memory Graph
> 3. WPTabBarController and other non-login related view controllers should be deallocated

Here the app's starting memory was 155 MB, went to 313 MB after logging in, and stayed high (323 MB) after logging out (the increase is explained by navigating to the log out screen). I saw you mention how there's many other retain cycles that are out of scope of this PR and that could explain this. I checked the Memory Graph and indeed `WPTabBarController` is no longer there (this was while testing the Jetpack app), which is a good sign.

This PR does appear to fix a retain cycle, but it might not be reducing the app's memory consumption overall. If so, putting this in the release notes as I suggested earlier might more sense once we've addressed some of the other retain cycles. What do you think?

---

It's not easy to grok how the window management works (especially across the two apps + the migration flow). However, I think the approach is promising. I'm happy to approve this once I can spend more time on the regression testing tomorrow and can get a bit more confidence in the fix. On a side note, splitting this into multiple PRs could be an option so that if for example a change needs to be reverted, we won't need to revert all these separate fixes.

staskus commented:
Thanks for the review, @guarani !

> This PR does appear to fix a retain cycle, but it might not be reducing the app's memory consumption overall. If so, putting this in the release notes as I suggested earlier might more sense once we've addressed some of the other retain cycles. What do you think?

Yes, you're right. I had the expectation of fixing these retain cycles to clean the memory but there're more issues that I will address with future PRs. I didn't want to make this one too big. I will update the release notes. 

> On a side note, splitting this into multiple PRs could be an option so that if for example a change needs to be reverted, we won't need to revert all these separate fixes.

A lot of the small retain cycle improvements include retain cycles with `BlogDashboardViewController`, so only fixing all of them change anything fundamentally. However, we could split `RootViewCoordinator` fixes and all the retain cycle fixes in 2 PRs if needed.

staskus committed:
Update RELEASE-NOTES.txt

staskus committed:
Merge branch 'trunk' into fix/21010-memory-not-deallocated-after-logging-out

guarani reviewed (approved):
Finished manual testing and everything is working as before.

Tested on iPhone:
- [x] Login & Logout (On both Jetpack and WordPress apps). Tested:
  - [x] Tested WP app
  - [x] Tested WP to JP app migration
  - [x] Tested JP app
- [x] Regression test: WordPress without static posters
- [x]  Regression test: WordPress with static posters
- [x] Regression test: Dashboard Blaze card

Tested on iPad:
- [x] Regression test: Dashboard Quick Actions
- [x] Regression test: Dashboard navigation 
- [x] Regression test: Reader navigation

The code changes look good and I think this is ready to merge 👍
